README.md
# acorn128 NIF

[![Build Status](https://travis-ci.org/potatosalad/erlang-acorn128.svg?branch=master)](https://travis-ci.org/potatosalad/erlang-acorn128) [![Hex.pm](https://img.shields.io/hexpm/v/acorn128.svg)](https://hex.pm/packages/acorn128)

Implementation of ACORN, second choice for use case 1 (Lightweight applications; resource constrained environments) from the [CAESAR competition final portfolio](https://competitions.cr.yp.to/caesar-submissions.html).

See [&ldquo;ACORN: A Lightweight Authenticated Cipher (v3)&rdquo;](https://competitions.cr.yp.to/round3/acornv3.pdf) by [Hongjun Wu](https://www3.ntu.edu.sg/home/wuhj/) for more information.

## Installation

Add `acorn128` to your project's dependencies in `mix.exs`

```elixir
defp deps do
  [
    {:acorn128, "~> 0.0.1"}
  ]
end
```

Add `acorn128` to your project's dependencies in your `Makefile` for [`erlang.mk`](https://github.com/ninenines/erlang.mk) or the following to your `rebar.config`

```erlang
{deps, [
  {acorn128, ".*", {git, "git://github.com/potatosalad/erlang-acorn128.git", {branch, "master"}}}
]}.
```

## Usage

### Simple API

_Erlang_

```erlang
%% Encryption
Key = crypto:strong_rand_bytes(16), % Key = <<33,126,75,53,220,99,55,188,229,97,154,128,147,91,14,249>>.
IV = crypto:strong_rand_bytes(16),  % IV  = <<60,52,165,219,131,60,192,154,183,168,77,231,240,130,245,118>>.
PlainText = <<"hello, world">>,
AAD = <<"test">>,
{CipherText, CipherTag} = acorn128:encrypt(Key, IV, PlainText, AAD).
% CipherText = <<3,105,56,55,26,140,36,60,63,214,228,107>>.
% CipherTag  = <<28,8,46,253,158,35,196,249,12,174,189,248,160,215,189,10>>.

%% Decryption
PlainText = acorn128:decrypt(Key, IV, CipherText, AAD, CipherTag).
% PlainText = <<"hello, world">>.
```

_Elixir_

```elixir
## Encryption
key = :crypto.strong_rand_bytes(16) # key = <<33,126,75,53,220,99,55,188,229,97,154,128,147,91,14,249>>
iv = :crypto.strong_rand_bytes(16)  # iv  = <<60,52,165,219,131,60,192,154,183,168,77,231,240,130,245,118>>
plaintext = "hello, world"
aad = "test"
{ciphertext, ciphertag} = :acorn128.encrypt(key, iv, plaintext, aad)
# ciphertext = <<3,105,56,55,26,140,36,60,63,214,228,107>>
# ciphertag  = <<28,8,46,253,158,35,196,249,12,174,189,248,160,215,189,10>>

## Decryption
^plaintext = :acorn128.decrypt(key, iv, ciphertext, aad, ciphertag)
# plaintext = "hello, world"
```

### New Crypto API

_Erlang_

```erlang
%% Encryption
Key = crypto:strong_rand_bytes(16), % Key = <<33,126,75,53,220,99,55,188,229,97,154,128,147,91,14,249>>.
IV = crypto:strong_rand_bytes(16),  % IV  = <<60,52,165,219,131,60,192,154,183,168,77,231,240,130,245,118>>.
PlainText = <<"hello, world">>,
AAD = <<"test">>,
{CipherText, CipherTag} = acorn128:crypto_one_time_aead(acorn128v3, Key, IV, PlainText, AAD, true).
% CipherText = <<3,105,56,55,26,140,36,60,63,214,228,107>>.
% CipherTag  = <<28,8,46,253,158,35,196,249,12,174,189,248,160,215,189,10>>.

%% Decryption
PlainText = acorn128:crypto_one_time_aead(acorn128v3, Key, IV, CipherText, AAD, CipherTag, false).
% PlainText = <<"hello, world">>.
```

_Elixir_

```elixir
## Encryption
key = :crypto.strong_rand_bytes(16) # key = <<33,126,75,53,220,99,55,188,229,97,154,128,147,91,14,249>>
iv = :crypto.strong_rand_bytes(16)  # iv  = <<60,52,165,219,131,60,192,154,183,168,77,231,240,130,245,118>>
plaintext = "hello, world"
aad = "test"
{ciphertext, ciphertag} = :acorn128.crypto_one_time_aead(:acorn128v3, key, iv, plaintext, aad, true)
# ciphertext = <<3,105,56,55,26,140,36,60,63,214,228,107>>
# ciphertag  = <<28,8,46,253,158,35,196,249,12,174,189,248,160,215,189,10>>

## Decryption
^plaintext = :acorn128.crypto_one_time_aead(:acorn128v3, key, iv, ciphertext, aad, ciphertag, false)
# plaintext = "hello, world"
```

### SUPERCOP API

_Erlang_

```erlang
%% Encryption
K = crypto:strong_rand_bytes(16),    % K    = <<33,126,75,53,220,99,55,188,229,97,154,128,147,91,14,249>>.
NPub = crypto:strong_rand_bytes(16), % NPub = <<60,52,165,219,131,60,192,154,183,168,77,231,240,130,245,118>>.
M = <<"hello, world">>,
AD = <<"test">>,
C = acorn128:crypto_aead_encrypt(M, AD, NPub, K).
% C = <<3,105,56,55,26,140,36,60,63,214,228,107,28,8,46,253,158,35,196,249,12,174,189,248,160,215,189,10>>.

%% Decryption
M = acorn128:crypto_aead_decrypt(C, AD, NPub, K).
% M = <<"hello, world">>.
```

_Elixir_

```elixir
## Encryption
k = :crypto.strong_rand_bytes(16)    # k    = <<33,126,75,53,220,99,55,188,229,97,154,128,147,91,14,249>>
npub = :crypto.strong_rand_bytes(16) # npub = <<60,52,165,219,131,60,192,154,183,168,77,231,240,130,245,118>>
m = "hello, world"
ad = "test"
c = :acorn128.crypto_aead_encrypt(m, ad, npub, k)
# c = <<3,105,56,55,26,140,36,60,63,214,228,107,28,8,46,253,158,35,196,249,12,174,189,248,160,215,189,10>>

## Decryption
^m = :acorn128.crypto_aead_decrypt(c, ad, npub, k)
# m = "hello, world"
```


CHANGELOG.md
# Changelog

## 0.0.1 (2019-12-24)

* Initial Release
* Support for `acorn128v3` based on SUPERCOP reference implementation.


test/acorn128_ct.erl
%% -*- mode: erlang; tab-width: 4; indent-tabs-mode: 1; st-rulers: [70] -*-
%% vim: ts=4 sw=4 ft=erlang noet
-module(acorn128_ct).

-include_lib("common_test/include/ct.hrl").

%% API
-export([start/2]).
-export([stop/1]).
-export([progress_start/0]).
-export([progress_stop/1]).

%% Internal API
-export([init/1]).

%% Macros
-define(RED, "\e[0;31m").
-define(GREEN, "\e[0;32m").
-define(YELLOW, "\e[0;33m").
-define(WHITE, "\e[0;37m").
-define(CYAN, "\e[0;36m").
-define(RESET, "\e[0m").
-define(TIME, 1000). % timer:seconds(1)

start(Group, Config) ->
	Now = os:timestamp(),
	io:format(user, "~s[  ] ~s :: ~s~s", [?WHITE, format_utc_timestamp(Now), Group, ?RESET]),
	{ok, Progress} = progress_start(),
	[{acorn128_ct, {Progress, Now}} | Config].

stop(Config) ->
	Now = os:timestamp(),
	{Progress, Old} = ?config(acorn128_ct, Config),
	ok = progress_stop(Progress),
	Diff = timer:now_diff(Now, Old),
	io:format(user, "~s[OK] ~s :: ~s elapsed~s~n", [?GREEN, format_utc_timestamp(Now), format_elapsed_time(Diff), ?RESET]),
	ok.

progress_start() ->
	Ref = erlang:make_ref(),
	{ok, Pid} = proc_lib:start(?MODULE, init, [{self(), Ref}]),
	{ok, {Ref, Pid}}.

progress_stop({Ref, Pid}) ->
	Pid ! {stop, self(), Ref},
	receive
		Ref ->
			ok
	after
		1000 ->
			ok
	end.

%% @private
format_elapsed_time(USec) ->
	Micro = USec rem 1000000,
	Second = ((USec - Micro) div 1000000) rem 60,
	Minute = ((USec - (Second * 1000000)) div 6000000) rem 60,
	Hour = ((USec - (Minute * 6000000)) div 360000000) rem 24,
	io_lib:format("~2..0w:~2..0w:~2..0w.~6..0w", [Hour,Minute,Second,Micro]).

%% @private
format_utc_timestamp(TS = {_, _, Micro}) ->
	{{Year,Month,Day},{Hour,Minute,Second}} = calendar:now_to_universal_time(TS),
	io_lib:format("~4..0w-~2..0w-~2..0w ~2..0w:~2..0w:~2..0w.~6..0w", [Year,Month,Day,Hour,Minute,Second,Micro]).

%% @private
init({Parent, Ref}) ->
	process_flag(trap_exit, true),
	ok = proc_lib:init_ack(Parent, {ok, self()}),
	{ok, TRef} = timer:send_interval(?TIME, {tick, Ref}),
	loop(Ref, TRef).

%% @private
loop(Ref, TRef) ->
	receive
		{tick, Ref} ->
			io:format(user, "~s.~s", [?WHITE, ?RESET]),
			loop(Ref, TRef);
		{stop, Parent, Ref} when is_pid(Parent) ->
			io:format(user, "~n", []),
			catch timer:cancel(TRef),
			Parent ! Ref,
			exit(normal);
		Info ->
			io:format(user, "~n~s[~s] received unhandled message:~n~p~s~n", [?RED, ?MODULE, Info, ?RESET]),
			loop(Ref, TRef)
	end.

src/acorn128.erl
%% -*- mode: erlang; tab-width: 4; indent-tabs-mode: 1; st-rulers: [70] -*-
%% vim: ts=4 sw=4 ft=erlang noet
-module(acorn128).

%% Simple API
-export([encrypt/4]).
-export([decrypt/5]).
%% New Crypto API
-export([crypto_one_time_aead/6]).
-export([crypto_one_time_aead/7]).
%% SUPERCOP API
-export([crypto_aead_encrypt/4]).
-export([crypto_aead_decrypt/4]).
%% Internal API
-export([priv_dir/0]).

%%%===================================================================
%%% Simple API Functions
%%%===================================================================

encrypt(Key, IV, PlainText, AAD) ->
	crypto_one_time_aead(acorn128v3, Key, IV, PlainText, AAD, true).

decrypt(Key, IV, CipherText, AAD, CipherTag) ->
	crypto_one_time_aead(acorn128v3, Key, IV, CipherText, AAD, CipherTag, false).

%%%===================================================================
%%% New Crypto API Functions
%%%===================================================================

crypto_one_time_aead(Cipher, Key, IV, InText, AAD, EncFlag) ->
	crypto_one_time_aead(Cipher, Key, IV, InText, AAD, 16, EncFlag).

crypto_one_time_aead(Cipher, Key, IV, InText, AAD, TagOrTagLength, EncFlag) ->
	acorn128_nif:crypto_one_time_aead(Cipher, Key, IV, InText, AAD, TagOrTagLength, EncFlag).

%%%===================================================================
%%% SUPERCOP API Functions
%%%===================================================================

crypto_aead_encrypt(M, AD, NPub, K) ->
	{CipherText, CipherTag} = encrypt(K, NPub, M, AD),
	<<CipherText/binary, CipherTag/binary>>.

crypto_aead_decrypt(C, AD, NPub, K) ->
	case byte_size(C) of
		CipherTextLen0 when CipherTextLen0 >= 16 ->
			CipherTextLen = CipherTextLen0 - 16,
			<<CipherText:CipherTextLen/binary, CipherTag/binary>> = C,
			decrypt(K, NPub, CipherText, AD, CipherTag);
		_ ->
			error
	end.

%%%===================================================================
%%% Internal API Functions
%%%===================================================================

-spec priv_dir() -> file:filename_all().
priv_dir() ->
	case code:priv_dir(?MODULE) of
		{error, bad_name} ->
			case code:which(?MODULE) of
				Filename when is_list(Filename) ->
					filename:join([filename:dirname(Filename), "../priv"]);
				_ ->
					"../priv"
			end;
		Dir ->
			Dir
	end.

%%%-------------------------------------------------------------------
%%% Internal functions
%%%-------------------------------------------------------------------


c_src/nif/acorn128_nif.c
// -*- mode: c++; tab-width: 4; indent-tabs-mode: nil; st-rulers: [132] -*-
// vim: ts=4 sw=4 ft=c++ et

#include "acorn128_nif.h"
#include "api.h"
#include "crypto_one_time_aead.h"

static ERL_NIF_TERM ATOM_acorn128v3;
static ERL_NIF_TERM ATOM_badarg;
static ERL_NIF_TERM ATOM_error;
static ERL_NIF_TERM ATOM_false;
static ERL_NIF_TERM ATOM_notsup;
static ERL_NIF_TERM ATOM_true;

/* All nif functions return a valid value or throws an exception */
#define EXCP(Env, Id, Str)                                                                                                         \
    enif_raise_exception((Env), enif_make_tuple3((Env), (Id),                                                                      \
                                                 enif_make_tuple2((Env), enif_make_string((Env), __FILE__, (ERL_NIF_LATIN1)),      \
                                                                  enif_make_int((Env), __LINE__)),                                 \
                                                 enif_make_string((Env), (Str), (ERL_NIF_LATIN1))))

#define EXCP_NOTSUP(Env, Str) EXCP((Env), ATOM_notsup, (Str))
#define EXCP_BADARG(Env, Str) EXCP((Env), ATOM_badarg, (Str))
#define EXCP_ERROR(Env, Str) EXCP((Env), ATOM_error, (Str))

/* NIF Function Declarations */

static ERL_NIF_TERM acorn128_nif_crypto_one_time_aead_7(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[]);

/* NIF Function Definitions */

/* acorn128_nif:crypto_one_time_aead/7 */

static ERL_NIF_TERM
acorn128_nif_crypto_one_time_aead_7(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
{
    int encflag;
    unsigned int tag_len;
    ErlNifBinary k;
    ErlNifBinary npub;
    ErlNifBinary in_text;
    ErlNifBinary ad;
    ErlNifBinary tag;
    ERL_NIF_TERM out_term;
    ERL_NIF_TERM tag_term;
    unsigned char *out = NULL;
    if (argc != 7) {
        return EXCP_NOTSUP(env, "argc must be 7");
    }
    /* Fetch the flag telling if we are going to encrypt (=true) or decrypt (=false) */
    if (argv[6] == ATOM_true) {
        encflag = 1;
    } else if (argv[6] == ATOM_false) {
        encflag = 0;
    } else {
        return EXCP_BADARG(env, "Bad EncFlag");
    }
    if (!enif_is_atom(env, argv[0])) {
        return EXCP_BADARG(env, "non-atom Cipher");
    }
    if (argv[0] != ATOM_acorn128v3) {
        return EXCP_NOTSUP(env, "Unsupported Cipher");
    }
    if (!enif_inspect_iolist_as_binary(env, argv[1], &k)) {
        return EXCP_BADARG(env, "non-binary Key");
    }
    if (k.size != CRYPTO_KEYBYTES) {
        return EXCP_BADARG(env, "Bad Key size");
    }
    if (!enif_inspect_iolist_as_binary(env, argv[2], &npub)) {
        return EXCP_BADARG(env, "non-binary IV");
    }
    if (npub.size != CRYPTO_NPUBBYTES) {
        return EXCP_BADARG(env, "Bad IV size");
    }
    if (!enif_inspect_iolist_as_binary(env, argv[3], &in_text)) {
        return EXCP_BADARG(env, "non-binary InText");
    }
    if (!enif_inspect_iolist_as_binary(env, argv[4], &ad)) {
        return EXCP_BADARG(env, "non-binary AAD");
    }
    if (encflag) {
        if (!enif_get_uint(env, argv[5], &tag_len)) {
            return EXCP_BADARG(env, "non-integer TagLength");
        }
        if (tag_len != CRYPTO_ABYTES) {
            return EXCP_BADARG(env, "Bad TagLength");
        }
        out = enif_make_new_binary(env, in_text.size, &out_term);
        if (out == NULL) {
            return EXCP_ERROR(env, "Can't allocate 'OutCryptoText' binary");
        }
        tag.data = enif_make_new_binary(env, CRYPTO_ABYTES, &tag_term);
        if (tag.data == NULL) {
            return EXCP_ERROR(env, "Can't allocate 'OutTag' binary");
        }
        (void)acorn128v3_crypto_one_time_aead_encrypt(out, tag.data, in_text.data, in_text.size, ad.data, ad.size, npub.data,
                                                      k.data);
        return enif_make_tuple2(env, out_term, tag_term);
    } else {
        if (!enif_inspect_iolist_as_binary(env, argv[5], &tag)) {
            return EXCP_BADARG(env, "non-binary Tag");
        }
        if (tag.size != CRYPTO_ABYTES) {
            return EXCP_BADARG(env, "Bad Tag size");
        }
        out = enif_make_new_binary(env, in_text.size, &out_term);
        if (out == NULL) {
            return EXCP_ERROR(env, "Can't allocate 'OutPlainText' binary");
        }
        if (acorn128v3_crypto_one_time_aead_decrypt(out, in_text.data, in_text.size, tag.data, ad.data, ad.size, npub.data,
                                                    k.data) == 0) {
            return out_term;
        } else {
            return ATOM_error;
        }
    }
}

/* NIF Callbacks */

static ErlNifFunc acorn128_nif_funcs[] = {
    {"crypto_one_time_aead", 7, acorn128_nif_crypto_one_time_aead_7, ERL_NIF_DIRTY_JOB_CPU_BOUND}};

static void acorn128_nif_make_atoms(ErlNifEnv *env);
static int acorn128_nif_load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info);
static int acorn128_nif_upgrade(ErlNifEnv *env, void **new_priv_data, void **old_priv_data, ERL_NIF_TERM load_info);
static void acorn128_nif_unload(ErlNifEnv *env, void *priv_data);

static void
acorn128_nif_make_atoms(ErlNifEnv *env)
{
#define ATOM(Id, Value)                                                                                                            \
    {                                                                                                                              \
        Id = enif_make_atom(env, Value);                                                                                           \
    }
    ATOM(ATOM_acorn128v3, "acorn128v3");
    ATOM(ATOM_badarg, "badarg");
    ATOM(ATOM_error, "error");
    ATOM(ATOM_false, "false");
    ATOM(ATOM_notsup, "notsup");
    ATOM(ATOM_true, "true");
#undef ATOM
}

static int
acorn128_nif_load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info)
{
    int retval = 0;

    // Unused
    (void)priv_data;
    (void)load_info;

    /* Initialize common atoms */
    (void)acorn128_nif_make_atoms(env);

    return retval;
}

static int
acorn128_nif_upgrade(ErlNifEnv *env, void **priv_data, void **old_priv_data, ERL_NIF_TERM load_info)
{
    int retval = 0;

    // Unused
    (void)priv_data;
    (void)old_priv_data;
    (void)load_info;

    /* Initialize common atoms */
    (void)acorn128_nif_make_atoms(env);

    return retval;
}

static void
acorn128_nif_unload(ErlNifEnv *env, void *priv_data)
{
    // Unused
    (void)env;
    (void)priv_data;
    return;
}

ERL_NIF_INIT(acorn128_nif, acorn128_nif_funcs, acorn128_nif_load, NULL, acorn128_nif_upgrade, acorn128_nif_unload)


c_src/nif/encrypt.c
#include <string.h>
#include "crypto_one_time_aead.h"

#define maj(x, y, z) (((x) & (y)) ^ ((x) & (z)) ^ ((y) & (z)))
#define ch(x, y, z) (((x) & (y)) ^ (((x) ^ 1) & (z)))

static unsigned char KSG128(unsigned char *state);
static unsigned char FBK128(unsigned char *state, unsigned char *ks, unsigned char ca, unsigned char cb);
static void Encrypt_StateUpdate128_1bit(unsigned char *state, unsigned char plaintextbit, unsigned char *ciphertextbit,
                                        unsigned char *ks, unsigned char ca, unsigned char cb);
static void Decrypt_StateUpdate128_1bit(unsigned char *state, unsigned char *plaintextbit, unsigned char ciphertextbit,
                                        unsigned char *ks, unsigned char ca, unsigned char cb);
static void acorn128_enc_onebyte(unsigned char *state, unsigned char plaintextbyte, unsigned char *ciphertextbyte,
                                 unsigned char *ksbyte, unsigned char cabyte, unsigned char cbbyte);
static void acorn128_dec_onebyte(unsigned char *state, unsigned char *plaintextbyte, unsigned char ciphertextbyte,
                                 unsigned char *ksbyte, unsigned char cabyte, unsigned char cbbyte);
static void acorn128_initialization(const unsigned char *key, const unsigned char *iv, unsigned char *state);
static void acorn128_tag_generation(unsigned long long msglen, unsigned long long adlen, unsigned char maclen, unsigned char *mac,
                                    unsigned char *state);
static void acorn128_memzero(void *const pnt, const size_t len);

static unsigned char
KSG128(unsigned char *state)
{
    return (state[12] ^ state[154] ^ maj(state[235], state[61], state[193]) ^ ch(state[230], state[111], state[66]));
}

static unsigned char
FBK128(unsigned char *state, unsigned char *ks, unsigned char ca, unsigned char cb)
{
    unsigned char f;
    *ks = KSG128(state);
    f = state[0] ^ (state[107] ^ 1) ^ maj(state[244], state[23], state[160]) ^ (ca & state[196]) ^ (cb & (*ks));
    return f;
}

// encrypt one bit
static void
Encrypt_StateUpdate128_1bit(unsigned char *state, unsigned char plaintextbit, unsigned char *ciphertextbit, unsigned char *ks,
                            unsigned char ca, unsigned char cb)
{
    unsigned int j;
    unsigned char f;

    state[289] ^= state[235] ^ state[230];
    state[230] ^= state[196] ^ state[193];
    state[193] ^= state[160] ^ state[154];
    state[154] ^= state[111] ^ state[107];
    state[107] ^= state[66] ^ state[61];
    state[61] ^= state[23] ^ state[0];

    f = FBK128(state, ks, ca, cb);

    for (j = 0; j <= 291; j++)
        state[j] = state[j + 1];
    state[292] = f ^ plaintextbit;
    *ciphertextbit = *ks ^ plaintextbit;
}

// decrypt one bit
static void
Decrypt_StateUpdate128_1bit(unsigned char *state, unsigned char *plaintextbit, unsigned char ciphertextbit, unsigned char *ks,
                            unsigned char ca, unsigned char cb)
{
    unsigned int j;
    unsigned char f;

    state[289] ^= state[235] ^ state[230];
    state[230] ^= state[196] ^ state[193];
    state[193] ^= state[160] ^ state[154];
    state[154] ^= state[111] ^ state[107];
    state[107] ^= state[66] ^ state[61];
    state[61] ^= state[23] ^ state[0];

    f = FBK128(state, ks, ca, cb);

    for (j = 0; j <= 291; j++)
        state[j] = state[j + 1];
    *plaintextbit = *ks ^ ciphertextbit;
    state[292] = f ^ *plaintextbit;
}

// encrypt one byte
static void
acorn128_enc_onebyte(unsigned char *state, unsigned char plaintextbyte, unsigned char *ciphertextbyte, unsigned char *ksbyte,
                     unsigned char cabyte, unsigned char cbbyte)
{
    // unsigned char i, t[4]; // t[4] is unused
    unsigned char i;
    unsigned char plaintextbit, ciphertextbit, kstem, ca, cb;

    *ciphertextbyte = 0;
    kstem = 0;
    *ksbyte = 0;
    for (i = 0; i < 8; i++) {
        ca = (cabyte >> i) & 1;
        cb = (cbbyte >> i) & 1;
        plaintextbit = (plaintextbyte >> i) & 1;
        Encrypt_StateUpdate128_1bit(state, plaintextbit, &ciphertextbit, &kstem, ca, cb);
        *ciphertextbyte |= (ciphertextbit << i);
        *ksbyte |= (kstem << i);
    }
}

// decrypt one byte
static void
acorn128_dec_onebyte(unsigned char *state, unsigned char *plaintextbyte, unsigned char ciphertextbyte, unsigned char *ksbyte,
                     unsigned char cabyte, unsigned char cbbyte)
{
    unsigned char i;
    unsigned char plaintextbit, ciphertextbit, ks, ca, cb;

    (void)ksbyte;

    *plaintextbyte = 0;
    for (i = 0; i < 8; i++) {
        ca = (cabyte >> i) & 1;
        cb = (cbbyte >> i) & 1;
        ciphertextbit = (ciphertextbyte >> i) & 1;
        Decrypt_StateUpdate128_1bit(state, &plaintextbit, ciphertextbit, &ks, ca, cb);
        *plaintextbyte |= (plaintextbit << i);
    }
}

// The initialization state of ACORN
/*The input to initialization is the 128-bit key; 128-bit IV;*/
static void
acorn128_initialization(const unsigned char *key, const unsigned char *iv, unsigned char *state)
{
    int i, j;
    unsigned char m[293], ks, tem;

    // initialize the state to 0
    for (j = 0; j <= 292; j++)
        state[j] = 0;

    // set the value of m
    for (j = 0; j <= 15; j++)
        m[j] = key[j];
    for (j = 16; j <= 31; j++)
        m[j] = iv[j - 16];
    for (j = 32; j <= 223; j++)
        m[j] = key[j & 0xf];
    m[32] ^= 1;

    // run the cipher for 1792 steps
    for (i = 0; i < 224; i++) {
        acorn128_enc_onebyte(state, m[i], &tem, &ks, 0xff, 0xff);
    }
}

// the finalization state of acorn
static void
acorn128_tag_generation(unsigned long long msglen, unsigned long long adlen, unsigned char maclen, unsigned char *mac,
                        unsigned char *state)
{
    int i;
    unsigned char plaintextbyte = 0;
    unsigned char ciphertextbyte = 0;
    unsigned char ksbyte = 0;

    // unused
    (void)msglen;
    (void)adlen;
    (void)maclen;

    for (i = 0; i < 768 / 8; i++) {
        acorn128_enc_onebyte(state, plaintextbyte, &ciphertextbyte, &ksbyte, 0xff, 0xff);
        if (i >= (768 / 8 - 16)) {
            mac[i - (768 / 8 - 16)] = ksbyte;
        }
    }
}

static void
acorn128_memzero(void *const pnt, const size_t len)
{
    volatile unsigned char *volatile pnt_ = (volatile unsigned char *volatile)pnt;
    size_t i = (size_t)0U;

    while (i < len) {
        pnt_[i++] = 0U;
    }
}

// encrypt a message
int
acorn128v3_crypto_one_time_aead_encrypt(unsigned char *c, unsigned char *tag, const unsigned char *m, unsigned long long mlen,
                                        const unsigned char *ad, unsigned long long adlen, const unsigned char *npub,
                                        const unsigned char *k)
{
    unsigned long long i;
    unsigned char plaintextbyte, ciphertextbyte, ksbyte;
    unsigned char state[293];
    unsigned char ca, cb;

    // initialization stage
    acorn128_initialization(k, npub, state);

    // process the associated data
    for (i = 0; i < adlen; i++) {
        acorn128_enc_onebyte(state, ad[i], &ciphertextbyte, &ksbyte, 0xff, 0xff);
    }

    for (i = 0; i < 256 / 8; i++) {
        if (i == 0)
            plaintextbyte = 0x1;
        else
            plaintextbyte = 0;

        if (i < 128 / 8)
            ca = 0xff;
        else
            ca = 0;

        cb = 0xff;

        acorn128_enc_onebyte(state, plaintextbyte, &ciphertextbyte, &ksbyte, ca, cb);
    }

    // process the plaintext
    for (i = 0; i < mlen; i++) {
        acorn128_enc_onebyte(state, m[i], &(c[i]), &ksbyte, 0xff, 0);
    }

    for (i = 0; i < 256 / 8; i++) {
        if (i == 0)
            plaintextbyte = 0x1;
        else
            plaintextbyte = 0;

        if (i < 128 / 8)
            ca = 0xff;
        else
            ca = 0;

        cb = 0;

        acorn128_enc_onebyte(state, plaintextbyte, &ciphertextbyte, &ksbyte, ca, cb);
    }

    // finalization stage, we assume that the tag length is a multiple of bytes
    acorn128_tag_generation(mlen, adlen, 16, tag, state);

    // Zero out all state
    acorn128_memzero(&i, sizeof(unsigned long long));
    acorn128_memzero(&plaintextbyte, sizeof(unsigned char));
    acorn128_memzero(&ciphertextbyte, sizeof(unsigned char));
    acorn128_memzero(&ksbyte, sizeof(unsigned char));
    acorn128_memzero(&state, 293 * sizeof(unsigned char));
    acorn128_memzero(&ca, sizeof(unsigned char));
    acorn128_memzero(&cb, sizeof(unsigned char));

    return 0;
}

// decrypt a message
int
acorn128v3_crypto_one_time_aead_decrypt(unsigned char *m, const unsigned char *c, unsigned long long clen, const unsigned char *tag,
                                        const unsigned char *ad, unsigned long long adlen, const unsigned char *npub,
                                        const unsigned char *k)
{
    unsigned long long i;
    unsigned char plaintextbyte, ciphertextbyte, ksbyte;
    unsigned char state[293];
    unsigned char challenge[16];
    unsigned char check = 0;
    unsigned char ca, cb;

    // initialization stage
    acorn128_initialization(k, npub, state);

    // process the associated data
    for (i = 0; i < adlen; i++) {
        acorn128_enc_onebyte(state, ad[i], &ciphertextbyte, &ksbyte, 0xff, 0xff);
    }

    for (i = 0; i < 256 / 8; i++) {
        if (i == 0)
            plaintextbyte = 0x1;
        else
            plaintextbyte = 0;

        if (i < 128 / 8)
            ca = 0xff;
        else
            ca = 0;

        cb = 0xff;

        acorn128_enc_onebyte(state, plaintextbyte, &ciphertextbyte, &ksbyte, ca, cb);
    }

    // process the ciphertext
    for (i = 0; i < clen; i++) {
        acorn128_dec_onebyte(state, &m[i], c[i], &ksbyte, 0xff, 0);
    }

    for (i = 0; i < 256 / 8; i++) {
        if (i == 0)
            plaintextbyte = 0x1;
        else
            plaintextbyte = 0;

        if (i < 128 / 8)
            ca = 0xff;
        else
            ca = 0;

        cb = 0;

        acorn128_enc_onebyte(state, plaintextbyte, &ciphertextbyte, &ksbyte, ca, cb);
    }

    // finalization stage, we assume that the tag length is a multiple of bytes
    acorn128_tag_generation(clen, adlen, 16, challenge, state);

    for (i = 0; i < 16; i++)
        check |= (challenge[i] ^ tag[i]);

    // Zero out all state
    acorn128_memzero(&i, sizeof(unsigned long long));
    acorn128_memzero(&challenge, 16 * sizeof(unsigned char));
    acorn128_memzero(&plaintextbyte, sizeof(unsigned char));
    acorn128_memzero(&ciphertextbyte, sizeof(unsigned char));
    acorn128_memzero(&ksbyte, sizeof(unsigned char));
    acorn128_memzero(&state, 293 * sizeof(unsigned char));
    acorn128_memzero(&ca, sizeof(unsigned char));
    acorn128_memzero(&cb, sizeof(unsigned char));

    if (check == 0) {
        return 0;
    } else {
        return -1;
    }
}
